### PR TITLE
Windows: Fullscreen

### PR DESCRIPTION
**Pull request by kazatsuyu**
~~This is still a work in progress, but to show the public where I am working on it~~ Ready for review

TODO:
- [x] Justify fullscreen size to display
- [x] Record and apply restored size

Release Notes:

- N/A

### CODE REVIEW

**Comment by kazatsuyu** [comment]
This PR enables full screen, but leaves a problem that is difficult to solve immediately, so I am writing it down.

When using Vulkan (and possibly other graphics APIs) on Windows, it automatically goes into "exclusive fullscreen" mode when a frameless window is matched to the display size. In this state, composition by the window manager will not work, for example, task switching by Alt+Tab will not work properly.

Instead, the application can draw without the delay caused by composition, which is useful for games, but not desirable for text editors where task switching is not possible.

It needs to be explicitly disabled using the `VK_EXT_full_screen_exclusive` extension, which is a future issue since it requires changes to the `blade`.

**Comment by JunkuiZhang** [comment]
I have some thoughts. Since zed was originally built for macOS, and the concept of fullscreen on macOS and Windows platforms is quite different. Based on my personal experience using macOS and Windows, going fullscreen with an application is a common operation on macOS, where I can switch between different applications by swiping left or right with four fingers. On the other hand, on Windows, applications generally only go fullscreen for games, while other applications just maximize the window. For example, on Windows, VSCode does not have a fullscreen option, whereas on macOS, going fullscreen is a common operation. Therefore, I think that on the Windows platform, implementing `toggle_fullscreen` as `toggle_maximized` might be more in line with the typical user interaction on Windows.

**Comment by kazatsuyu** [comment]
VSCode on Windows has a fullscreen mode. F11 key is assigned to it by default, same as Zed. When using touch pad, three finger swipe can switch applications even if it is fullscreen.
Zed has `zoom` command to maximize window. Although Zed has no `unzoom` command AFAIK, and it is useful to implement this, I think there is no reason to omit fullscreen mode.

File: crates/gpui/src/platform/windows/window.rs [hunk 1]
```diff
@@ -1709,3 +1800,21 @@ fn logical_point(x: f32, y: f32, scale_factor: f32) -> Point<Pixels> {
...
+enum FullScreenState {
+    None,
+    Normal {
+        style: WINDOW_STYLE,
+        x: i32,
+        y: i32,
+        cx: i32,
+        cy: i32,
+    },
+    Maximized(WINDOW_STYLE),
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:1813]: Why is maximized included here?

File: crates/gpui/src/platform/windows/window.rs [hunk 2]
```diff
@@ -125,17 +127,98 @@ impl WindowsWindowInner {
...
         unsafe { IsIconic(self.hwnd) }.as_bool()
     }
 
+    fn is_fullscreen(&self) -> bool {
+        self.fullscreen.borrow().is_fullscreen()
+    }
+
+    async fn toggle_fullscreen(self: Rc<Self>) {
+        let mut state = self.fullscreen.borrow_mut();
+        let (next_state, style, x, y, cx, cy, swp_flags) = match &*state {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/window.rs:148]: What's going on with this code? Does windows not restore the window size when disabling fullscreen?

**Comment by mikayla-maki** [comment]
Excited to have this :)

File: crates/gpui/src/platform/windows/window.rs (hunk 1 above)
Comment (kazatsuyu) [reply crates/gpui/src/platform/windows/window.rs]: I thought it was necessary to separate the cases of going full screen from the normal window and the case of  the maximized one, but I found that the same process was fine, so I removed it.

File: crates/gpui/src/platform/windows/window.rs (hunk 2 above)
Comment (kazatsuyu) [reply crates/gpui/src/platform/windows/window.rs]: Windows does not have an explicit fullscreen state; a frameless window that matches the display size is treated as fullscreen. Fullscreening is done by moving the window, so the size must be saved to restore it.

Comment (mikayla-maki) [reply crates/gpui/src/platform/windows/window.rs]: Ahhh, I see.

**Pull request merged**